### PR TITLE
Extend ruff_dev formatter script to compute statistics and format a project

### PR DESCRIPTION
**Pull request by konstin**
## Summary

This extends the `ruff_dev` formatter script util. Instead of only doing stability checks, you can now choose different compatible options on the CLI and get statistics.

* It adds an option the formats all files that ruff would check to allow looking at an entire black-formatted repository with `git diff` 
* It computes the [Jaccard index](https://en.wikipedia.org/wiki/Jaccard_index) as a measure of deviation between input and output, which is useful as single number metric for assessing our current deviations from black.
* It adds progress bars to both the single projects as well as the multi-project mode.
* It adds an option to write the multi-project output to a file

Sample usage:

```
$ cargo run --bin ruff_dev -- format-dev --stability-check crates/ruff/resources/test/cpython
$ cargo run --bin ruff_dev -- format-dev --stability-check /home/konsti/projects/django
Syntax error in /home/konsti/projects/django/tests/test_runner_apps/tagged/tests_syntax_error.py: source contains syntax errors (parser error): BaseError { error: UnrecognizedToken(Name { name: "syntax_error" }, None), offset: 131, source_path: "<filename>" }
Found 0 stability errors in 2755 files (jaccard index 0.911) in 9.75s
$ cargo run --bin ruff_dev -- format-dev --write /home/konsti/projects/django
```

Options:

```
Several utils related to the formatter which can be run on one or more repositories. The selected set of files in a repository is the same as for `ruff check`.

* Check formatter stability: Format a repository twice and ensure that it looks that the first and second formatting look the same. * Format: Format the files in a repository to be able to check them with `git diff` * Statistics: The subcommand the Jaccard index between the (assumed to be black formatted) input and the ruff formatted output

Usage: ruff_dev format-dev [OPTIONS] [FILES]...

Arguments:
  [FILES]...
          Like `ruff check`'s files. See `--multi-project` if you want to format an ecosystem checkout

Options:
      --stability-check
          Check stability
          
          We want to ensure that once formatted content stays the same when formatted again, which is known as formatter stability or formatter idempotency, and that the formatter prints syntactically valid code. As our test cases cover only a limited amount of code, this allows checking entire repositories.

      --write
          Format the files. Without this flag, the python files are not modified

      --format <FORMAT>
          Control the verbosity of the output
          
          [default: default]

          Possible values:
          - minimal: Filenames only
          - default: Filenames and reduced diff
          - full:    Full diff and invalid code

  -x, --exit-first-error
          Print only the first error and exit, `-x` is same as pytest

      --multi-project
          Checks each project inside a directory, useful e.g. if you want to check all of the ecosystem checkouts

      --error-file <ERROR_FILE>
          Write all errors to this file in addition to stdout. Only used in multi-project mode
```

## Test Plan

I ran this on django (2755 files, jaccard index 0.911) and discovered a magic trailing comma problem and that we really needed to implement import formatting. I ran the script on cpython to identify https://github.com/astral-sh/ruff/pull/5558.

### CODE REVIEW

**Comment by konstin** [comment]
Current dependencies on/for this PR:
* main
  * **PR #5491** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5491" target="_blank">Graphite</a> 
    * **PR #5492** <a href="https://app.graphite.dev/github/pr/astral-sh/ruff/5492" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/astral-sh/ruff/5492?utm_source=stack-comment).

File: .cargo/config.toml [hunk 1]
```diff
@@ -12,6 +12,7 @@ rustflags = [
   "-Wclippy::pedantic",
   # Allowed pedantic lints
   "-Wclippy::char_lit_as_u8",
+  "-Aclippy::cast_precision_loss",
```
Comment (MichaReiser) [inline .cargo/config.toml:15]: I suggest moving this into a dedicated PR. I found the rule useful so far (I've been liberal with allowing the rule in methods where I knew it isn't a problem)

File: crates/ruff_cli/src/args.rs [hunk 2]
```diff
@@ -74,7 +74,8 @@ pub enum Command {
     },
 }
 
-#[derive(Clone, Debug, clap::Args)]
+// The `Parser` derive is for ruff_dev, for ruff_cli `Args` would be sufficient
+#[derive(Clone, Debug, clap::Parser)]
```
Comment (MichaReiser) [inline crates/ruff_cli/src/args.rs:78]: Does this change the size of the resulting binary? If so, move behind a feature flag? Why does the existing `WrapperArgs` approach no longer work?

File: crates/ruff_dev/src/format_project.rs [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+        JaccardAccumulator {
+            black_input: self.black_input + rhs.black_input,
+            ruff_output: self.ruff_output + rhs.ruff_output,
+            intersection: self.intersection + rhs.intersection,
+        }
+    }
+}
+
+pub(crate) fn main(args: &Args) -> anyhow::Result<()> {
+    let start = Instant::now();
```
Comment (MichaReiser) [inline crates/ruff_dev/src/format_project.rs:68]: What's your motivation for creating another command vs integrating it into the `check_formatter_stability`? My main concern is that we start duplicating the code for each command (we probably want to have an option to run the script over multiple projects in the future...)

File: crates/ruff_dev/src/format_project.rs [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+                ChangeTag::Equal => jaccard_accumulator.intersection += 1,
+            }
+        }
+        // Simple atomic write.
+        // The file is in a directory so it must have a parent. Surprisingly `DirEntry` doesn't
+        // give us access without unwrap
+        let mut file = NamedTempFile::new_in(input_path.parent().unwrap())?;
+        file.write_all(reformatted.as_code().as_bytes())?;
+        // "If a file exists at the target path, persist will atomically replace it."
+        file.persist(input_path)?;
```
Comment (MichaReiser) [inline crates/ruff_dev/src/format_project.rs:192]: Why does this file write any output? 

Rome had a useful script. It allowed you to generate a markdown report for all prettier test. The report contained:

* For files where the formatting differs
	* The Jacard index
	* The formatting diff
* The Jacard index for the whole project

I found this report extremely useful because:
* I had syntax highlighting of the patch
* I can see which files have the biggest differences
* I get an overview metric

**Review by MichaReiser — COMMENTED** [review]
How much effort would it be to unify the `checek_formatter_stability` and `format_project` commands? I'm worried that we're having too many CLIs (the one in `ruff_python_formatter`, `ruff format`, `check_stability`, `format_project`).

File: crates/ruff_cli/src/args.rs (hunk 2 above)
Comment (konstin) [reply crates/ruff_cli/src/args.rs:78]: `WrapperArgs` was mostly a hack around clap's limitations, i think this is a much cleaner solution.

For `ruff-0.0.277-py3-none-manylinux_2_34_x86_64.whl` from `CARGO_TARGET_DIR=target-maturin maturin build --release --strip -j 2`:
* On main: 5593346
* This branch: 5504378

Surprisingly even better than main, so i'd say the impact on binary size is in the realm of noise.

File: crates/ruff_python_formatter/src/lib.rs [hunk 5]
```diff
@@ -84,16 +91,40 @@ where
     }
 }
 
-pub fn format_module(contents: &str, options: PyFormatOptions) -> Result<Printed> {
+#[derive(Error, Debug)]
+pub enum FormatModuleError {
```
Comment (konstin) [inline crates/ruff_python_formatter/src/lib.rs:95]: the `anyhow::Result` from `format_node` had irked me already and now i had the motivation to change it

**Comment by konstin** [comment]
I have to check the details with fresh eyes, but the command works and does the right thing

**Review by MichaReiser — APPROVED** [review]
Looks code wise. You may want to consider writing a report instead of formatting the files, to avoid overlap with the functionality of ruff and ruff python formatter

**Comment by konstin** [comment]
> You may want to consider writing a report instead of formatting the files, to avoid overlap with the functionality of ruff and ruff python formatter

Open to that, just not a priority for me atm
